testing/src/test/java/io/stargate/it/storage/StargateContainer.java

```diff
@@ -0,0 +1,538 @@
+            }
+          }
+        };
+
+    Runtime.getRuntime().addShutdownHook(containerShutdown);
+  }
+
+  private static Container container() {
+    if (container == null) {
+      throw new IllegalStateException("Stargate container has not been configured");
+    }
+
+    return container;
+  }
+
+  private static void enter(AtomicBoolean flag) {
+    if (!flag.compareAndSet(false, true)) {
+      throw new IllegalStateException(
+          String.format(
+              "Concurrent execution with %s is not supported",
+              StargateContainer.class.getSimpleName()));
+    }
+  }
+
+  private static void leave(AtomicBoolean flag) {
+    flag.compareAndSet(true, false);
+  }
+
+  private static StargateParameters parameters(StargateSpec spec, ExtensionContext context)
+      throws Exception {
+    Builder builder = StargateParameters.builder();
+
+    String customizer = spec.parametersCustomizer().trim();
+    if (!customizer.isEmpty()) {
+      Object testInstance = context.getTestInstance().orElse(null);
+      Class<?> testClass = context.getRequiredTestClass();
+      Method method = testClass.getMethod(customizer, Builder.class);
+      method.invoke(testInstance, builder);
+    }
+
+    return builder.build();
+  }
+
+  private static synchronized void start(ExtensionContext context) throws Exception {
+    ClusterConnectionInfo backend = ExternalStorage.connectionInfo();
+
+    StargateSpec spec =
+        context
+            .getElement()
+            .flatMap(e -> Optional.ofNullable(e.getAnnotation(StargateSpec.class)))
+            .orElseGet(
+                () ->
+                    context
+                        .getTestClass()
+                        .flatMap(c -> Optional.ofNullable(c.getAnnotation(StargateSpec.class)))
+                        .orElse(DefaultSpecHolder.get()));
+
+    StargateParameters params = parameters(spec, context);
+
+    Container c = container;
+    if (c != null) {
+      if (!c.matches(backend, spec, params)) {
+        c.stop();
+      } else {
+        // reuse container
+        LOG.info("Reusing Stargate container for the next test.");
+        return;
+      }
+    }
+
+    c = new Container(backend, spec, params);
+    c.start();
+    container = c;
+  }
+
+  @Override
+  public void beforeAll(ExtensionContext context) throws Exception {
+    enter(suiteExecuting);
+    start(context);
+  }
+
+  @Override
+  public void afterAll(ExtensionContext context) {
+    leave(suiteExecuting);
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) throws Exception {
+    enter(testExecuting);
+    start(context);
+  }
+
+  @Override
+  public void afterEach(ExtensionContext context) throws Exception {
+    leave(testExecuting);
+  }
+
+  private boolean isStargateConnectionInfo(ParameterContext parameterContext) {
+    return parameterContext.getParameter().getType().isAssignableFrom(StargateConnectionInfo.class);
```

Comment:
I think that's too coarse-grained, see #188.

Reply:
fixed